src/main/scala/metadata.scala

```diff
@@ -0,0 +1,83 @@
+package suffuse
+
+/** AKey is an attribute key: a Metadata map key which carries
+ *  both the type of the corresponding value, and a value to
+ *  provide when there is no value in the map. This as opposed
+ *  to wrapping everything in sight in Option.
+ */
+final class AKey[A](description: String, empty: => A) {
+  def emptyValue: A = empty
+  override def toString = description
+}
+
+/** Attr is a dependently typed key/value pair.
+ */
+trait Attr extends Any {
+  type Type
+  def key: AKey[Type]
+  def value: Type
+
+  def pair                               = key -> value
+  def matches[A](implicit that: AKey[A]) = key == that
+  def to_s                               = s"$key: $value"
+}
+
+/** The simple implementation of Attr, which ties the knot between the
+ *  user-facing type parameter and the type member which couples key and value.
+ */
+final case class AttrOf[A](key: AKey[A], value: A) extends Attr {
```

Comment:
I do not think this one should be public. You might want the type alias `type AttrOf[A] = Attr { type Type = A }`